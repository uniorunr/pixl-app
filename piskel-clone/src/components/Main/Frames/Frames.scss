.frames-section {
  display: flex;
  justify-content: flex-start;
  flex-direction: column;
  align-items: center;
  align-self: flex-start;
  max-height: 90%;
  min-width: 105px;
  margin: 0 20px;
  padding: 5px;
  font-weight: 700;
  font-size: 1.2rem;
  background-color: white;
  border: 2px solid black;
  cursor: context-menu;
  overflow-y: scroll;

  .frames-section__add-frame-button {
    margin: 20px auto;
    font-size: 1.4rem;
    cursor: pointer;
    background: transparent;
    border: none;
    outline: none;

    span {
      margin-left: 5px;
      font-family: "Roboto", "Arial", sans-serif;
      font-weight: 700;
    }
  }
}
