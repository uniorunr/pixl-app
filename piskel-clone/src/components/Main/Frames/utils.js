const setActiveFrame = (activeFrame, canvasItems, index) => {
  let finalIndex = null;
  if (index === activeFrame) {
    finalIndex = 0;
  } else if (index === canvasItems.length - 1) {
    finalIndex = activeFrame;
  } else if (
    index === canvasItems.length - 2
    && activeFrame === canvasItems.length - 1
  ) {
    finalIndex = index;
  } else if (activeFrame === canvasItems.length - 1) {
    finalIndex = canvasItems.length - 2;
  } else {
    finalIndex = activeFrame;
  }
  return finalIndex;
};

export default setActiveFrame;
