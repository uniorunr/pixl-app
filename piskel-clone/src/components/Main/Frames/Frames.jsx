import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Frame from './Frame/Frame';
import './Frames.scss';
import { setActiveFrame, translateActiveFrame } from './utils';

class Frames extends Component {
  state = {
    canvasItems: [],
    frameKeys: [0],
    activeFrame: 0,
    duplicate: null,
  };

  componentDidMount = () => {
    this.setState({
      canvasItems: [
        <Frame
          index={0}
          key="0"
          makeActive={this.makeActive}
          removeFrame={this.removeFrame}
          duplicateFrame={this.duplicateFrame}
          resetDuplicate={this.resetDuplicate}
        />,
      ],
    });
  };

  makeActive = (index) => {
    this.setState({
      activeFrame: index,
    });
  };

  resetDuplicate = () => {
    this.setState({
      duplicate: null,
    });
  };

  addFrame = () => {
    const { canvasItems, frameKeys } = this.state;

    this.setState({
      canvasItems: [
        ...canvasItems,
        <Frame
          index={Math.max(...frameKeys) + 1}
          key={Math.max(...frameKeys) + 1}
          removeFrame={this.removeFrame}
          makeActive={this.makeActive}
          duplicateFrame={this.duplicateFrame}
          resetDuplicate={this.resetDuplicate}
        />,
      ],
      frameKeys: [...frameKeys, Math.max(...frameKeys) + 1],
      activeFrame: frameKeys.length,
    });

    const mainCanvas = document.querySelector('#canvas');
    const mainCtx = mainCanvas.getContext('2d');
    mainCtx.clearRect(0, 0, mainCanvas.width, mainCanvas.height);
  };

  removeFrame = (index) => {
    const { canvasItems, frameKeys, activeFrame } = this.state;
    const indexToTranslate = setActiveFrame(activeFrame, canvasItems, index);
    canvasItems.splice(index, 1);
    frameKeys.splice(index, 1);
    translateActiveFrame(indexToTranslate);

    this.setState({
      canvasItems: [...canvasItems],
      frameKeys: [...frameKeys],
      activeFrame:
        indexToTranslate > canvasItems.length - 1
          ? canvasItems.length - 1
          : indexToTranslate,
    });
  };

  duplicateFrame = (index) => {
    const { canvasItems, frameKeys } = this.state;
    const { framesArray } = this.props;

    const frameForInsert = (
      <Frame
        index={index + 1}
        key={Math.max(...frameKeys) + 1}
        removeFrame={this.removeFrame}
        makeActive={this.makeActive}
        duplicateFrame={this.duplicateFrame}
        resetDuplicate={this.resetDuplicate}
        contextSource={framesArray[index]}
      />
    );
    canvasItems.splice(index + 1, 0, frameForInsert);
    frameKeys.splice(index + 1, 0, Math.max(...frameKeys) + 1);

    this.setState({
      canvasItems: [...canvasItems],
      frameKeys: [...frameKeys],
      duplicate: index + 1,
      activeFrame: index + 1,
    });
  };

  render() {
    const {
      canvasItems, frameKeys, activeFrame, duplicate,
    } = this.state;
    const { framesArray } = this.props;
    const keys = canvasItems.map(item => item.key);

    return (
      <section className="frames-section">
        {keys.map((item, index) => (
          <Frame
            index={index}
            key={frameKeys[index]}
            removeFrame={this.removeFrame}
            makeActive={this.makeActive}
            duplicateFrame={this.duplicateFrame}
            resetDuplicate={this.resetDuplicate}
            active={index === activeFrame}
            duplicate={index === duplicate}
            framesArray={framesArray}
          />
        ))}
        <button
          type="button"
          className="frames-section__add-frame-button"
          onClick={this.addFrame}
        >
          <i className="fas fa-plus" />
          <span>Add</span>
        </button>
      </section>
    );
  }
}

Frames.propTypes = {
  framesArray: PropTypes.instanceOf(Array),
};

Frames.defaultProps = {
  framesArray: [],
};

export default Frames;
