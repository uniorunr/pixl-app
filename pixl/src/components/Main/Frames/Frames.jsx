import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { DragDropContext, Droppable, Draggable } from 'react-beautiful-dnd';
import Frame from './Frame/Frame';
import List from './ListWrapper/ListWrapper';
import * as actions from '../../../actions/actions';
import { setActiveFrame, translateActiveFrame } from './utils';
import './Frames.scss';

class Frames extends Component {
  componentDidMount() {
    document.addEventListener('keydown', ({ code, shiftKey }) => {
      const { framesShortcuts, activeFrame } = this.props;
      if (shiftKey) {
        if (code === framesShortcuts.duplicate.shortcut) {
          this.duplicateFrame(activeFrame);
        } else if (
          code === framesShortcuts.remove.shortcut
          && activeFrame !== 0
        ) {
          this.removeFrame(activeFrame);
        }
      }
    });
  }

  componentDidUpdate(prevProps) {
    const {
      framesData,
      framesArray,
      layerKeys,
      activeLayer,
      frameKeys,
      activeFrame,
      updateFramesData,
    } = this.props;
    if (
      JSON.stringify(prevProps.framesData) !== JSON.stringify(framesData)
      || JSON.stringify(prevProps.frameKeys) !== JSON.stringify(frameKeys)
      || JSON.stringify(prevProps.layerKeys) !== JSON.stringify(layerKeys)
    ) {
      const layerKey = `layer${layerKeys[activeLayer]}`;
      if (layerKeys[activeLayer] >= 0) {
        const tempFramesData = { ...framesData };
        tempFramesData[layerKey] = framesArray.map(item => item.toDataURL());
        updateFramesData(tempFramesData);
        sessionStorage.setItem('framesData', JSON.stringify(tempFramesData));
      }
      sessionStorage.setItem('activeFrame', `${activeFrame}`);
      sessionStorage.setItem('frameKeys', JSON.stringify(frameKeys));
    }
  }

  duplicateFrame = (index) => {
    const {
      frameKeys,
      updateDuplicateFrameIndex,
      updateActiveFrameIndex,
      updateFrameKeys,
      framesArray,
    } = this.props;
    const tempFrameKeys = [...frameKeys];
    tempFrameKeys.splice(index + 1, 0, Math.max(...tempFrameKeys) + 1);
    updateFrameKeys([...tempFrameKeys]);
    updateActiveFrameIndex(index + 1);
    const origin = framesArray[index];
    const target = framesArray[index + 1];
    const context = target.getContext('2d');
    context.imageSmoothingEnabled = false;
    context.drawImage(origin, 0, 0, target.width, target.height);
    translateActiveFrame(index);
    updateDuplicateFrameIndex(null);
    sessionStorage.setItem('activeFrame', index + 1);
  };

  removeFrame = (index) => {
    const {
      frameKeys,
      activeFrame,
      updateActiveFrameIndex,
      updateFrameKeys,
      framesData,
      activeLayer,
      layerKeys,
      updateFramesData,
    } = this.props;
    const indexToTranslate = setActiveFrame(activeFrame, frameKeys, index);
    const tempFrameKeys = [...frameKeys];
    const tempFramesData = { ...framesData };
    const newActiveFrameIndex = indexToTranslate > tempFrameKeys.length - 2
      ? tempFrameKeys.length - 2
      : indexToTranslate;
    const layerKey = `layer${layerKeys[activeLayer]}`;
    tempFrameKeys.splice(index, 1);
    tempFramesData[layerKey].splice(index, 1);
    updateFramesData(tempFramesData);
    updateFrameKeys([...tempFrameKeys]);
    updateActiveFrameIndex(newActiveFrameIndex);
    translateActiveFrame(indexToTranslate);
    sessionStorage.setItem('activeFrame', newActiveFrameIndex);
  };

  addFrame = () => {
    const {
      frameKeys,
      canvas,
      updateActiveFrameIndex,
      updateFrameKeys,
    } = this.props;

    const newFrameKeys = [...frameKeys, Math.max(...frameKeys) + 1];
    updateFrameKeys(newFrameKeys);
    updateActiveFrameIndex(frameKeys.length);
    sessionStorage.setItem('frameKeys', JSON.stringify(newFrameKeys));
    const mainCtx = canvas.getContext('2d');
    mainCtx.clearRect(0, 0, canvas.width, canvas.height);
  };

  onDragEnd = (result) => {
    const { destination, source } = result;
    const {
      frameKeys,
      activeFrame,
      updateActiveFrameIndex,
      updateFrameKeys,
      framesArray,
      framesData,
      activeLayer,
      layerKeys,
      updateFramesData,
    } = this.props;
    const destIndex = destination ? destination.index : null;
    const srcIndex = source.index;
    const activeFrameKey = frameKeys[activeFrame];
    if (destination && destIndex !== srcIndex) {
      const tempFrameKeys = [...frameKeys];
      const srcItem = tempFrameKeys[srcIndex];
      tempFrameKeys.splice(srcIndex, 1);
      tempFrameKeys.splice(destIndex, 0, srcItem);
      updateActiveFrameIndex(tempFrameKeys.indexOf(activeFrameKey));
      updateFrameKeys([...tempFrameKeys]);

      const tempFramesData = { ...framesData };
      const layerKey = `layer${layerKeys[activeLayer]}`;
      tempFramesData[layerKey] = framesArray.map(item => item.toDataURL());
      updateFramesData(tempFramesData);
      sessionStorage.setItem('framesData', JSON.stringify(tempFramesData));
    }
  };

  render() {
    const { frameKeys, activeFrame, duplicateIndex } = this.props;

    return (
      <section className="frames-section">
        <DragDropContext onDragEnd={this.onDragEnd}>
          <Droppable droppableId="frames">
            {provided => (
              <List provided={provided} innerRef={provided.innerRef}>
                {frameKeys.map((item, index) => (
                  <Draggable
                    draggableId={`item${item}`}
                    index={index}
                    key={item}
                  >
                    {provided1 => (
                      <Frame
                        index={index}
                        key={item}
                        removeFrame={this.removeFrame}
                        active={index === activeFrame}
                        duplicate={index === duplicateIndex}
                        frameKeys={frameKeys}
                        provided={provided1}
                        innerRef={provided1.innerRef}
                      />
                    )}
                  </Draggable>
                ))}
                {provided.placeholder}
              </List>
            )}
          </Droppable>
        </DragDropContext>
        <button
          type="button"
          className="frames-section__add-frame-button"
          onClick={this.addFrame}
        >
          <i className="fas fa-plus" />
          <span>Add</span>
        </button>
      </section>
    );
  }
}

Frames.propTypes = {
  framesArray: PropTypes.instanceOf(Array).isRequired,
  framesData: PropTypes.instanceOf(Object).isRequired,
  layerKeys: PropTypes.instanceOf(Array).isRequired,
  activeLayer: PropTypes.number.isRequired,
  framesShortcuts: PropTypes.instanceOf(Object).isRequired,
  frameKeys: PropTypes.instanceOf(Array).isRequired,
  activeFrame: PropTypes.number.isRequired,
  updateActiveFrameIndex: PropTypes.func.isRequired,
  updateFrameKeys: PropTypes.func.isRequired,
  canvas: PropTypes.instanceOf(Object),
  duplicateIndex: PropTypes.number,
  updateFramesData: PropTypes.func.isRequired,
  updateDuplicateFrameIndex: PropTypes.func.isRequired,
};

Frames.defaultProps = {
  canvas: null,
  duplicateIndex: null,
};

const mapStateToProps = state => ({
  framesShortcuts: state.frames.framesShortcuts,
  framesData: state.frames.framesData,
  frameKeys: state.frames.frameKeys,
  framesArray: state.frames.framesArray,
  activeFrame: state.frames.activeFrame,
  duplicateIndex: state.frames.duplicateIndex,
  layerKeys: state.layers.layerKeys,
  activeLayer: state.layers.activeLayer,
  canvas: state.canvas.canvasRef,
});

const mapDispatchToProps = (dispatch) => {
  const {
    updateActiveFrameIndex,
    updateFrameKeys,
    updateFramesData,
    updateDuplicateFrameIndex,
  } = bindActionCreators(actions, dispatch);
  return {
    updateActiveFrameIndex: (index) => {
      updateActiveFrameIndex(index);
    },
    updateFrameKeys: (keys) => {
      updateFrameKeys(keys);
    },
    updateFramesData: (data) => {
      updateFramesData(data);
    },
    updateDuplicateFrameIndex: (index) => {
      updateDuplicateFrameIndex(index);
    },
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Frames);
