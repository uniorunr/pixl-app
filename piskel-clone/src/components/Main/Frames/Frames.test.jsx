import React from 'react';
import renderer from 'react-test-renderer';
import Frames from './Frames';

function createNodeMock(element) {
  if (typeof element.type === 'string' && element.type === 'div') {
    const elm = document.createElement(element.type);
    elm.setAttribute('data-react-beautiful-dnd-drag-handle', '0');
    return elm;
  }
  if (element.type === 'input') {
    return {
      current: {
        value: 23,
      },
    };
  }
  return null;
}

it('renders correctly', () => {
  const options = { createNodeMock };
  const tree = renderer.create(<Frames />, options).toJSON();
  expect(tree).toMatchSnapshot();
});
