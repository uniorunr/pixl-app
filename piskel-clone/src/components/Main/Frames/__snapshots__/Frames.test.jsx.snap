// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<section
  className="frames-section"
>
  <div
    className="frame frame0"
    data-active={true}
    data-id={0}
    role="presentation"
  >
    <canvas
      className="frame__canvas frame__canvas_active"
      height="128"
      id="frame0"
      onClick={[Function]}
      width="128"
    />
    <span
      className="frame__index"
    >
      1
    </span>
    <button
      className="frame__duplicate"
      onClick={[Function]}
      type="button"
    >
      <i
        className="fas fa-copy"
      />
    </button>
  </div>
  <button
    className="frames-section__add-frame-button"
    onClick={[Function]}
    type="button"
  >
    <i
      className="fas fa-plus"
    />
    <span>
      Add
    </span>
  </button>
</section>
`;
